### PR TITLE
Refactors PDA cartridge permissions

### PR DESCRIPTION
**Pull request by ohnopigeons**
Instead of an ungodly amount of vars they are now bit flags with an ungodly amount of #define's.

Also I renamed the detonate var and fixed a thing where the virus sending PDA will no longer return success when the target PDAs aren't detonatable in the first place.

### CODE REVIEW

File: code/game/objects/items/devices/PDA/virus_cart.dm [hunk 1]
```diff
@@ -66,15 +66,11 @@
...
 		var/difficulty = 0
 		if(target.cartridge)
-			difficulty += target.cartridge.access_medical
-			difficulty += target.cartridge.access_security
-			difficulty += target.cartridge.access_engine
-			difficulty += target.cartridge.access_clown
-			difficulty += target.cartridge.access_janitor
-			difficulty += target.cartridge.access_manifest * 2
+			difficulty += BitCount(target.cartridge.access&(CART_MEDICAL|CART_SECURITY|CART_ENGINE|CART_CLOWN|CART_JANITOR|CART_MANIFEST))
+			if(target.cartridge.access&CART_MANIFEST) difficulty++ //if cartridge has manifest access it has extra snowflake difficulty
```
Comment (lzimann) [inline code/game/objects/items/devices/PDA/virus_cart.dm]: new line for the if

Comment (ohnopigeons) [inline code/game/objects/items/devices/PDA/virus_cart.dm]: Before the if?

Comment (lzimann) [inline code/game/objects/items/devices/PDA/virus_cart.dm]: the `difficulty++` should be in a new line

Comment (ohnopigeons) [inline code/game/objects/items/devices/PDA/virus_cart.dm]: https://medium.com/@jonathanabrams/single-line-if-statements-2565c62ff492

Comment (lzimann) [inline code/game/objects/items/devices/PDA/virus_cart.dm]: https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#control-statements

Comment (ohnopigeons) [inline code/game/objects/items/devices/PDA/virus_cart.dm]: That is dumb but okay

Comment (lzimann) [inline code/game/objects/items/devices/PDA/virus_cart.dm]: What is dumb is linking an article about languages that use brackets as an excuse for byond.

Comment (ohnopigeons) [inline code/game/objects/items/devices/PDA/virus_cart.dm]: For something as simple as an iteration the all-in-one-line being more readable point still stands.

Comment (lzimann) [inline code/game/objects/items/devices/PDA/virus_cart.dm]: I don't agree, all in one line is pretty bad for readability, also looks really bad.

File: code/game/objects/items/devices/PDA/PDA.dm [hunk 2]
```diff
@@ -42,7 +42,7 @@ GLOBAL_LIST_EMPTY(PDAs)
 	var/notehtml = ""
 	var/notescanned = 0 // True if what is in the notekeeper was from a paper.
 	var/cart = "" //A place to stick cartridge menu information
-	var/detonate = 1 // Can the PDA be blown up?
+	var/detonatable = 1 // Can the PDA be blown up?
```
Comment (Cyberboss) [inline code/game/objects/items/devices/PDA/PDA.dm]: TRUE

File: code/game/objects/items/devices/PDA/PDA.dm [hunk 3]
```diff
@@ -153,31 +153,31 @@ GLOBAL_LIST_EMPTY(PDAs)
 				dat += "<li><a href='byond://?src=\ref[src];choice=2'><img src=pda_mail.png> Messenger</a></li>"
 
 				if (cartridge)
-					if (cartridge.access_clown)
+					if (cartridge.access&CART_CLOWN)
```
Comment (Cyberboss) [inline code/game/objects/items/devices/PDA/PDA.dm]: Add spaces between `access` `&` and the flag for each of these

File: code/game/objects/items/devices/PDA/PDA_types.dm [hunk 4]
```diff
@@ -23,7 +23,7 @@
 	icon_state = "NONE"
 	ttone = "data"
 	fon = 0
-	detonate = 0
+	detonatable = 0
```
Comment (Cyberboss) [inline code/game/objects/items/devices/PDA/PDA_types.dm]: FALSE

File: code/game/objects/items/devices/PDA/PDA_types.dm [hunk 5]
```diff
@@ -121,7 +121,7 @@
 	default_cartridge = /obj/item/weapon/cartridge/captain
 	inserted_item = /obj/item/weapon/pen/fountain/captain
 	icon_state = "pda-captain"
-	detonate = 0
+	detonatable = 0
```
Comment (Cyberboss) [inline code/game/objects/items/devices/PDA/PDA_types.dm]: FALSE

File: code/game/objects/items/devices/PDA/virus_cart.dm [hunk 6]
```diff
@@ -66,15 +66,12 @@
...
 		var/difficulty = 0
 		if(target.cartridge)
-			difficulty += target.cartridge.access_medical
-			difficulty += target.cartridge.access_security
-			difficulty += target.cartridge.access_engine
-			difficulty += target.cartridge.access_clown
-			difficulty += target.cartridge.access_janitor
-			difficulty += target.cartridge.access_manifest * 2
+			difficulty += BitCount(target.cartridge.access&(CART_MEDICAL|CART_SECURITY|CART_ENGINE|CART_CLOWN|CART_JANITOR|CART_MANIFEST))
+			if(target.cartridge.access&CART_MANIFEST) //if cartridge has manifest access it has extra snowflake difficulty
```
Comment (Cyberboss) [inline code/game/objects/items/devices/PDA/virus_cart.dm]: Spaces again

**Review by Cyberboss — DISMISSED** [review]

File: code/game/objects/items/devices/PDA/PDA.dm (hunk 3 above)
Comment (ohnopigeons) [inline code/game/objects/items/devices/PDA/PDA.dm]: https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#operators-and-spaces

File: code/game/objects/items/devices/PDA/virus_cart.dm (hunk 6 above)
Comment (ohnopigeons) [inline code/game/objects/items/devices/PDA/virus_cart.dm]: https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#operators-and-spaces

File: code/game/objects/items/devices/PDA/virus_cart.dm (hunk 1 above)
Comment (ohnopigeons) [inline code/game/objects/items/devices/PDA/virus_cart.dm]: `if(target.cartridge.access&CART_MANIFEST) difficulty++`
reads like a sentence in natural language, splitting them up is as awkward as a paragraph indent midsentence

File: code/game/objects/items/devices/PDA/PDA.dm (hunk 3 above)
Comment (optimumtact) [inline code/game/objects/items/devices/PDA/PDA.dm]: spaces > nonspaces

File: code/__HELPERS/unsorted.dm [hunk 7]
```diff
@@ -1414,3 +1414,10 @@ GLOBAL_PROTECT(valid_HTTPSGet)
 		mob_occupant = brain.brainmob
 
 	return mob_occupant
+
+//counts the number of bits in Byond's 16-bit width field
+//in constant time and memory!
+/proc/BitCount(bitfield)
```
Comment (optimumtact) [inline code/__HELPERS/unsorted.dm]: you could make this a macro if you were cool and awesome

File: code/game/objects/items/devices/PDA/virus_cart.dm (hunk 1 above)
Comment (lzimann) [inline code/game/objects/items/devices/PDA/virus_cart.dm]: Maybe a mistake, but it returned back to same line.

**Review by lzimann — CHANGES_REQUESTED** [review]

File: code/__HELPERS/unsorted.dm (hunk 7 above)
Comment (ohnopigeons) [inline code/__HELPERS/unsorted.dm]: `#define meme(B)	((B) - (((B)>>1) & 46811) - (((B)>>2) & 37449))`
`#define BITCOUNT(B)	((((meme(B)) + ((meme(B))>>3)) & 29127) % 63)`

Comment (optimumtact) [inline code/__HELPERS/unsorted.dm]: (B)ig guy

**Review by lzimann — APPROVED** [review]

**Pull request merged**